text/0000-relax-const-restrictions.md

```diff
@@ -0,0 +1,74 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This RFC allows `const fn` to exhibit different behavior during constant evaluation and runtime. While such a difference is often undesirable, it is not considered to be undefined behavior.
+
+If a `const fn` is able to detect whether it has been called during constant evaluation or at runtime (either through an intrinsic or a future language feature), then it is allowed to exhibit different behavior. Also, a `const fn` called at runtime can do anything a normal function can do, with no additional restrictions applied to it. It could open a file, call a system randomness API or gracefully exit the program. 
+
+At the time of writing, there is no way for a function to detect whether it was called at runtime or during constant evaluation in stable Rust and this RFC is not concerned with adding any, but it encourages future RFCs for adding this capability.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This RFC introduces a new concept called the "constness-context" of an execution of a function. A `const fn` is executed in a constant evaluation constness-context if it was called in a context where only `const fn` are allowed (like `static` initializers) or inside another `const fn` that was executed in a const constness-context. All other calls to a `const fn` (for example in `main`) are in a runtime constness-context.
+
+A `const fn` is now allowed to exhibit different behavior depending on it being called in a constant evaluation or runtime constness-context. Language features and standard library functions may also differ in behavior depending on the constness-context they have been used or called in.
+
+The constness-context of a function is therefore observable behavior.
+
+If a `const fn` is called in a runtime constness-context, no additional restrictions are applied to it, and it may do anything a non-`const fn` can (for example, calling into FFI).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+There are two main drawbacks to this RFC.
+
+Firstly, pure `const fn` under the old rules can be seen as a simple optimization opportunity for naive optimizers, as they could just reuse constant evaluation for `const fn` if the argument is known at compile time, even if the function is in a runtime context. This RFC makes such an optimization impossible. This is not seen as a problem by the author, as a more advanced optimizer (like LLVM) is able to remove these calls at compile time through means other than Rust's constant evaluation (inlining and constant folding). Also, a constant evaluation system can still evaluate executions in a runtime constness-context, as long as it behaves exactly like runtime.
```

Comment:
more concretely: the optimizer can run a purity checker that isn't bound by back-compat issues or other language concerns. So it can look a the bodies of functions and mark them as pure without changing the body of a (const or not) function becoming a breaking change for a library.